clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java

```diff
@@ -325,47 +330,54 @@ private DeepRecordsIterator(AbstractLegacyRecordBatch wrapperEntry, boolean ensu
 
-            long wrapperRecordOffset = wrapperEntry.lastOffset();
-            long wrapperRecordTimestamp = wrapperRecord.timestamp();
-            this.batches = new ArrayDeque<>();
+            long lastOffsetFromWrapper = wrapperEntry.lastOffset();
+            long timestampFromWrapper = wrapperRecord.timestamp();
+            this.innerEntries = new ArrayDeque<>();
 
             // If relative offset is used, we need to decompress the entire message first to compute
             // the absolute offset. For simplicity and because it's a format that is on its way out, we
             // do the same for message format version 0
             try {
                 while (true) {
-                    AbstractLegacyRecordBatch batch = logStream.nextBatch();
-                    if (batch == null)
+                    AbstractLegacyRecordBatch innerEntry = logStream.nextBatch();
+                    if (innerEntry == null)
                         break;
 
-                    LegacyRecord record = batch.outerRecord();
+                    LegacyRecord record = innerEntry.outerRecord();
                     byte magic = record.magic();
 
                     if (ensureMatchingMagic && magic != wrapperMagic)
                         throw new InvalidRecordException("Compressed message magic " + magic +
                                 " does not match wrapper magic " + wrapperMagic);
 
-                    if (magic > RecordBatch.MAGIC_VALUE_V0) {
+                    if (magic == RecordBatch.MAGIC_VALUE_V1) {
                         LegacyRecord recordWithTimestamp = new LegacyRecord(
                                 record.buffer(),
-                                wrapperRecordTimestamp,
+                                timestampFromWrapper,
                                 wrapperRecord.timestampType());
-                        batch = new BasicLegacyRecordBatch(batch.lastOffset(), recordWithTimestamp);
+                        innerEntry = new BasicLegacyRecordBatch(innerEntry.lastOffset(), recordWithTimestamp);
                     }
-                    batches.addLast(batch);
 
-                    // break early if we reach the last offset in the batch
-                    if (batch.offset() == wrapperRecordOffset)
-                        break;
+                    innerEntries.addLast(innerEntry);
                 }
 
-                if (batches.isEmpty())
+                if (innerEntries.isEmpty())
                     throw new InvalidRecordException("Found invalid compressed record set with no inner records");
 
-                if (wrapperMagic > RecordBatch.MAGIC_VALUE_V0)
-                    this.absoluteBaseOffset = wrapperRecordOffset - batches.getLast().lastOffset();
-                else
+                if (wrapperMagic == RecordBatch.MAGIC_VALUE_V1) {
+                    if (lastOffsetFromWrapper == 0) {
+                        // The outer offset may be 0 if this is produce data from an older client.
```

Comment:
Nit: instead of `older`, maybe we should say `certain versions of librdkafka`. The Java client never does this for `v1` as far as I know.